### PR TITLE
Add this_thread_executors

### PR DESCRIPTION
**Pull request by hkaiser**
- also add executor_traits<>:;has_pending_closures()
- refactoring interface of scheduling_loop
- exit scheduling loop without delay if called from inner scheduler (fixes performance regression)
- add missing #includes

### CODE REVIEW

**Comment by dcbdan** [comment]
What is the rationale behind has_pending_closures?

**Comment by hkaiser** [comment]
> What is the rationale behind has_pending_closures?

We essentially have 2 types of executors: stateless and stateful executors (I'm not sure if these terms are appropriate, though). Stateless executors run the tasks directly, while stateful executors employ the services of more complex scheduling engines to make sure the tasks are run. The `has_pending_closures` is most useful for stateful executors allowing to detect whether there are still tasks running, i.e. to detect whether it is safe to destroy the executor instance.

File: hpx/parallel/executors/executor_traits.hpp [hunk 1]
```diff
@@ -312,6 +312,29 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v3)
...
             return os_thread_count_helper::call(0, exec);
         }
+
+        ///////////////////////////////////////////////////////////////////////
+        struct has_pending_closures_helper
+        {
+            template <typename Executor>
+            static auto call(wrap_int, Executor& exec) -> bool
+            {
+                return hpx::get_os_thread_count();
```
Comment (sithhell) [inline hpx/parallel/executors/executor_traits.hpp]: Are the pending closures really the number of OS threads? wouldn't this rather return `false` for stateless executors?

**Comment by sithhell** [comment]
Is the purpose of this executor to confine all scheduled work on the current  OS thread? If yes, wouldn't a simple, single threaded scheduler be enough?

File: hpx/parallel/executors/this_thread_executors.hpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+#endif
+    {
+        /// Creates a new local_queue_executor
+        ///
+        /// \param max_punits   [in] The maximum number of processing units to
+        ///                     associate with the newly created executor.
+        /// \param min_punits   [in] The minimum number of processing units to
+        ///                     associate with the newly created executor
+        ///                     (default: 1).
+        ///
```
Comment (sithhell) [inline hpx/parallel/executors/this_thread_executors.hpp]: This looks like a copy and paste error.

File: hpx/parallel/executors/this_thread_executors.hpp [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+#endif
+    {
+        /// Creates a new static_priority_queue_executor
+        ///
+        /// \param max_punits   [in] The maximum number of processing units to
+        ///                     associate with the newly created executor.
+        /// \param min_punits   [in] The minimum number of processing units to
+        ///                     associate with the newly created executor
+        ///                     (default: 1).
+        ///
```
Comment (sithhell) [inline hpx/parallel/executors/this_thread_executors.hpp]: Same here.

File: hpx/parallel/executors/executor_traits.hpp (hunk 1 above)
Comment (hkaiser) [inline hpx/parallel/executors/executor_traits.hpp]: Good catch, that's a bug (copy&paste problem).

File: hpx/parallel/executors/this_thread_executors.hpp (hunk 2 above)
Comment (hkaiser) [inline hpx/parallel/executors/this_thread_executors.hpp]: Right, forgot to update the comments

**Comment by hkaiser** [comment]
> Is the purpose of this executor to confine all scheduled work on the current OS thread? If yes, wouldn't a simple, single threaded scheduler be enough?

Well, this is a single threaded scheduler guaranteeing that all work is executed on the same thread as it is initially executed. I needed this for HPXrx (https://github.com/STEllAR-GROUP/hpxrx), I think I can completely move it over to that project, it does not _have_ to go into the main HPX repo. OTOH, it might be useful in other contexts as well.

**Comment by sithhell** [comment]
On 07/21/2015 03:33 PM, Hartmut Kaiser wrote:
>> Is the purpose of this executor to confine all scheduled work on the current OS thread? If yes, wouldn't a simple, single threaded scheduler be enough?
> 
> Well, this is a single threaded scheduler guaranteeing that all work is executed on the same thread as it is initially executed. I needed this for HPXrx (https://github.com/STEllAR-GROUP/hpxrx), I think I can completely move it over to that project, it does not _have_ to go into the main HPX repo. OTOH, it might be useful in other contexts as well.

I am not saying it shouldn't go into the main HPX repo, it could indeed
be useful, what I am trying to get at is that it seems overkill for a
scheduler which is only used on a single core to include all the
atomics, lockfree algorithms and locks etc..
If I understand correctly, you never enqueue new work from more than one
thread and never steal any work so this might be a valuable optimization
which could come in the future.
With that being said, do you think we absolutely need multiple
schedulers for this executor?

> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/STEllAR-GROUP/hpx/pull/1669#issuecomment-123304995
> 


-- 
Thomas Heller
Friedrich-Alexander-Universität Erlangen-Nürnberg
Department Informatik - Lehrstuhl Rechnerarchitektur
Martensstr. 3
91058 Erlangen
Tel.: 09131/85-27018
Fax:  09131/85-27912
Email: thomas.heller@cs.fau.de

**Comment by hkaiser** [comment]
> If I understand correctly, you never enqueue new work from more than one thread and never steal any work so this might be a valuable optimization which could come in the future.

Right, it's not done anywhere currently, but should be possible to do.

> With that being said, do you think we absolutely need multiple schedulers for this executor?

Do you mean the static_priority and local_queue variants I added?  The static_priority one does not do work-stealing but enables priorities, while the local one does not expose priorities. We don't have a static local scheduler at this point, otherwise I would have used it instead.

**Comment by hkaiser** [comment]
All comments have been addressed. I also added a new static scheduler (no priorities and no work stealing) to reduce the overheads for this use case.

**Pull request merged**